text/0000-native-class-constructor-update.md

```diff
@@ -0,0 +1,220 @@
+- Start Date: 2018-06-14
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Native Class Constructor Update
+
+## Summary
+
+Update the behavior of EmberObject's constructor to defer object
+initialization.
+
+## Motivation
+
+Using native class syntax with EmberObject has almost reached full feature
+parity, meaning soon we'll be able to ship native classes and begin recommending
+them. This will do wonders for the Ember learning story, and will bring us in
+line with the wider Javascript community.
+
+However, early adopters of native classes have experienced some serious
+ergonomic issues due to the current behavior of the class constructor. The issue
+is caused by the fact that properties passed to `EmberObject.create` are
+assigned to the instance in the root class `constructor`. Due to the way that
+native class fields work, this means that they are assigned _before_ any
+subclasses' fields are assigned, causing subclass fields to overwrite any value
+passed to `create`:
+
+```js
+class Foo extends EmberObject {
+  bar = 'baz';
+}
+
+let foo = Foo.create({ bar: 'something different' });
+
+console.log(foo.bar); // 'baz'
+```
+
+This has made adoption very difficult, and is a consistent stumbling block for
+new users of native class syntax in Ember. Worse yet, it makes writing a codemod
+for converting to native class syntax very difficult because we don't have a
+clear target.
+
+For instance, given the above class, how would we convert the class field? Let's
+go through the various options:
+
+```js
+class Foo extends EmberObject {
+  // Does not work, for the reasons described above
+  bar = 'baz';
+
+  // Does not cover all cases. If we did `Foo.create({ bar: false })` it would
+  // still assign the default.
+  bar = this.bar || 'baz';
+
+  // This works, but is very verbose and not ideal
+  bar = this.hasOwnProperty('bar') ? this.bar : 'baz';
+
+  // This is one of the community accepted solutions, but it requires lodash
+  bar = _.defaultTo(this.bar, 'baz');
+
+  // This is another community accepted solution, but it requires
+  // @ember-decorators/argument, which is a separate library
+  @argument foo = 'bar';
+}
+```
+
+None of these is ideal. Instead, we can change the behavior of the constructor
+and the `create` method to circumvent this issue.
+
+This change _would_ be a breaking change to the behavior of native classes
+today, and a change from the previous class RFC. This will impact early adopters
+and should be made with that in mind. However, because native classes never
+officially shipped as part of Ember's public API (an announcement was not made,
+docs have not been written, etc), this RFC proposes that the change would _not_
+be considered a breaking change _for the purposes of semver_. This would allow
+us to ship the change during the Ember v3 release cycle, and prevent more code
+from being built on top of the previous behavior.
```

Comment:
It's worth note that the design landed on here *does not break* the behavior of any of the fallback solutions the community has had to embrace to deal with the pain point.

As such, this is a case where it's a "breaking change" in the strictest sense but is *extremely* unlikely to break any real-world code, because the real-world code has necessitated the use of one of those workarounds.